### PR TITLE
Replaces persistent collections with immutable collections

### PR DESCRIPTION
**Pull request by popematt**
**Issue #, if available:**

None. However, an offline discussion with someone who was trying out IonElement indicated that loading all values as IonElement was allocating up to 4x as memory as compared to loading the equivalent data as IonValue.

**Description of changes:**

This replaces `PersistentList` and `PersistentMap` with custom `ImmutableList` and `ImmutableMap` implementations. These implementations are _much_ simpler than `PersistentList` and `PersistentMap`—they can wrap any existing `List`/`Map` with very little overhead. I chose to do this because a lightweight wrapper will require less allocation than having to copy the data to a completely new instance of a `List` or `Map`. I think my implementations are pretty much the most minimal implementation one _could_ make for an immutable wrapper of `List` or `Map`.

I've also added an `ArrayBackedImmutableList` so that we don't need to allocate to convert the `String[]` of annotations into a `List` before wrapping it to make an `ImmutableList`.

Finally, I've added an `IonLocationBackedImmutableMap`—a special-purpose map implementation that can only contain one key-value pair, and the key can only be `$ion_location`. This saves a huge amount of allocation and significantly speeds up reading with location metadata to the point where reading `IonElement` with the additional overhead of location data is faster than reading `IonValue` (which has no location metadata).

**Compatibility**

The public API only exposes `List` and `Map`. `PersistentList` and `PersistentMap` have never been part of the API, so it is allowable to replace them with a different implementation.

However, Java users can still call mutating methods on a Kotlin `List`, so I did some manual testing to confirm that the behavior has not changed in that case. Both `PersistentList` and my `ImmutableList` implementations will throw `UnsupportedOperationException` if a method such as `add()` is called. The `PersistentMap` and my `ImmutableMap` implementations are the same.

<details>

<summary><b>Performance Data</b></summary>

Composition of some arbitrary data (~5kb)

API       | Time (ms/op)    | gc.alloc.rate.norm (B/op)
:---------|----------------:|-------------------:
IonValue  | `0.053 ± 0.001` |  `81128.034 ± 0.004`
ion-element@v1.2.0  | `0.047 ± 0.001` | `215416.030 ± 0.004`
#100  | `0.032 ± 0.001` | `131720.020 ± 0.003`
#101 (this PR) | `0.024 ± 0.001` |  `51952.015 ± 0.002`

Sample binary Ion application logs (~20MB)

API            | Time (ms/op)      | gc.alloc.rate.norm (B/op)
:--------------|------------------:|-----------------------------:
IonValue       | `635.450 ±  75.384` |  `625408257.218 ±  49.211`
ion-element@v1.2.0    | `630.640 ± 215.099` | `2348935671.047 ± 146.942`
#100 | `452.451 ± 283.069` | `2245469209.217 ± 113.536`
#101 (this PR) |  `312.335 ±  24.048` |  `446810105.670 ±  14.961`

Sample product catalog data (~40kb)

API            | Time (ms/op)    | gc.alloc.rate.norm (B/op) 
:--------------|--------------:|-----------------------:
IonValue       | `0.616 ± 0.093` |  `696584.433 ± 0.297`
ion-element@v1.2.0     | `0.559 ± 0.028` | `2999688.359 ± 0.046` 
#100 | `0.513 ± 0.007` | `2854072.329 ± 0.041` 
#101 (this PR) | `0.286 ± 0.016` |  `631800.182 ± 0.021`

For two of the samples, this PR has a normalized allocation rate that is about 1/4th of the current rate. For the one sample that displayed less of an improvement, it might be because it does not have as many lists/structs as the other data or it could be that it has a higher incidence of annotations. Either way, I'm not too concerned. Cumulatively, this change and #100 have improved the normalized memory allocation rate for all these samples from being ~4x more than `IonValue` to being less than `IonValue` (and in some cases, significantly less).

</details>

<details>

<summary>
<b>Performance Data (with IonLocation metadata)</b>
</summary>

_`(*)` indicates performance without location because location data is not supported by that API_

Composition of some arbitrary data (~5kb)

API       | Time (ms/op)    | gc.alloc.rate.norm (B/op)
:---------|----------------:|-------------------:
IonValue  | `(*) 0.053 ± 0.001` |  `(*) 81128.034 ± 0.004`
ion-element@v1.2.0  | `0.129 ± 0.003` | `714456.083 ± 0.011`
#100  | ` 0.076 ± 0.001` | `368472.048 ± 0.006`
#101 (this PR) | `0.033 ± 0.001` |    `113096.021 ± 0.002`

Sample binary Ion application logs (~20MB)

API            | Time (ms/op)      | gc.alloc.rate.norm (B/op)
:--------------|------------------:|-----------------------------:
IonValue       | `(*) 635.450 ±  75.384` |  `(*) 625408257.218 ±  49.211`
ion-element@v1.2.0   | `2492.157 ± 460.771` | `8622042546.800 ± 685.087`
#100 | `1296.107 ± 182.448` | `3979394547.314 ± 175.082`
#101 (this PR) | `453.957 ±  62.876` |    `679669196.884 ±  42.721`

Sample product catalog data (~40kb)

API            | Time (ms/op)    | gc.alloc.rate.norm (B/op) 
:--------------|--------------:|-----------------------:
IonValue       | `(*) 0.616 ± 0.093` |  `(*) 696584.433 ± 0.297` 
ion-element@v1.2.0     | `1.806 ± 0.055` | `10717049.154 ± 0.136`
#100 | `0.905 ± 0.025` | `4688840.633 ± 0.452` 
#101 (this PR) | ` 0.326 ± 0.006` |    `1026104.209 ± 0.027`

This PR brings a ~75% decrease in the allocation rate and a ~60% decrease in time per op compared to the previous commit, and cumulatively, the changes from 1.2.0 to this PR, provide up to a 90% decrease in the allocation rate, and up to a 80% decrease in time per op.

When compared to no location metadata, there is still a performance penalty associated with populating the location metadata, but while it was once 150-300% more time per op, it is now only 10-50% more time per op.

Populating the location metadata obviously requires more memory allocation than not doing it, so including `IonValue` here is not necessarily fair, but it is worth noting that for all of the sample data, the time per op is now ~40% less than IonValue—even with the location metadata.

</details>

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license._

### CODE REVIEW

File: src/main/kotlin/com/amazon/ionelement/impl/IonElementLoaderImpl.kt [hunk 1]
```diff
@@ -89,18 +89,13 @@ internal class IonElementLoaderImpl(private val options: IonElementLoaderOptions
...
-                        else -> emptyMetaContainer()
-                    }
-                }
-                else -> emptyMetaContainer()
-            }.toPersistentMap()
+            var metas = EMPTY_METAS
+            if (options.includeLocationMeta) {
+                val location = ionReader.currentLocation()
+                if (location != null) metas = location.toMetaContainer()
+            }
```
Comment (popematt) [inline src/main/kotlin/com/amazon/ionelement/impl/IonElementLoaderImpl.kt:98]: 🗺️ This is partly a stylistic change. The `toMetaContainer()` wraps the `IonLocation` in an `ImmutableMap` implementation, which is the main performance improvement for the location metadata. The rest of this change is stylistic to try to make it more readable and concise.

File: src/main/kotlin/com/amazon/ionelement/impl/IonElementLoaderImpl.kt [hunk 2]
```diff
@@ -89,18 +89,13 @@ internal class IonElementLoaderImpl(private val options: IonElementLoaderOptions
         return handleReaderException(ionReader) {
             val valueType = requireNotNull(ionReader.type) { "The IonReader was not positioned at an element." }
 
-            val annotations = ionReader.typeAnnotations!!.asList().toEmptyOrPersistentList()
+            val annotations = ionReader.typeAnnotations!!.toImmutableListUnsafe()
```
Comment (popematt) [inline src/main/kotlin/com/amazon/ionelement/impl/IonElementLoaderImpl.kt:92]: 🗺️ Previously, this line was creating a list from an array, and then creating a persistent list by copying all of the elements from the intermediate list (an `O(n)` operation). Now, it just wraps the array with an `ArrayBackedImmutableList` (an `O(1)` operation) in addition to `ArrayBackedImmutableList` having a smaller object layout to begin with.

File: src/main/kotlin/com/amazon/ionelement/impl/StructElementImpl.kt [hunk 3]
```diff
@@ -61,8 +58,9 @@ internal class StructElementImpl(
                 fieldsByNameBackingField =
                     fields
                         .groupBy { it.name }
-                        .map { structFieldGroup -> structFieldGroup.key to structFieldGroup.value.mapToEmptyOrPersistentList { it.value } }
-                        .toMap().toPersistentMap()
+                        .map { structFieldGroup -> structFieldGroup.key to structFieldGroup.value.map { it.value }.toImmutableListUnsafe() }
+                        .toMap()
+                        .toImmutableMapUnsafe()
```
Comment (popematt) [inline src/main/kotlin/com/amazon/ionelement/impl/StructElementImpl.kt:63]: 🗺️ I wanted to create only a minimal set of extension functions, so I didn't create `Iterable<K, V>.toImmutableMap()`. Instead, I just use the equivalent `.toMap().toImmutableMapUnsafe()` here.

**Review by popematt — COMMENTED** [review]
This change looks large, but ~18 of the modified files are just changing `Persistent*` to `Immutable*`.

I recommend starting with [ImmutableList.kt](https://github.com/amazon-ion/ion-element-kotlin/pull/101/files#diff-e796cb87f16971cbac7184a588ada52c25c301712d01ff453d3c84d114790b8d) and continuing down to `ImmutableMap.kt` and the associated tests for those two files. Then, take a look at [IonElementLoaderImpl](https://github.com/amazon-ion/ion-element-kotlin/pull/101/files#diff-bba24654a4bda5d9d4fdc0d7b683578dd905ee56ac6b04e51e0e0b10eb8deba1), and then look at the rest in whatever order you like.

**Review by jobarr-amzn — APPROVED** [review]

**Comment by popematt** [comment]
I've had a few offline questions that can be roughly summarized as "Why are the persistent collections so bad for performance?" I don't think persistent collections are bad, per se, but they might not be the right choice here.

1. Persistent collections make a tradeoff. They have a larger initial overhead in order that they might amortize the `O(n)` space for the elements when you start performing copy-on-write operations with that collection. Larger initialization cost means more costly to load `IonElement`s.
2. Just like any other immutable collection library, `kotlinx.collections.immutable` has to treat all input as untrusted input (i.e. make defensive copies of data) if it is going to correctly implement its immutability guarantees. In this PR, I am taking advantage of the fact that some of the underlying data, while mutable, cannot be leaked into a context where it would be mutated. This allows us to completely eliminate all defensive copies in `IonElementLoaderImpl`.

**Comment by popematt** [comment]
Well, I just looked again, and `kotlinx.collections.immutables.ImmutableList`, etc. are open interfaces _and_ there are adapter classes provided. I checked the performance, and it is essentially the same as what I've got, so there's a decision to be made.

Is it more valuable to minimize dependencies on other libraries or to use the interface of another library?

**Comment by jobarr-amzn** [comment]
If we *can* use `kotlinx.collections.immutables.ImmutableList`, etc. then I'd rather do that. Not because we don't have a perfectly good immutable list implementation in this PR, but because I'd rather own less code. 

Do we still need or benefit from our own specialized Ion location map implementation?

**Comment by popematt** [comment]
> If we _can_ use `kotlinx.collections.immutables.ImmutableList`, etc. then I'd rather do that. Not because we don't have a perfectly good immutable list implementation in this PR, but because I'd rather own less code.
> 
> Do we still need or benefit from our own specialized Ion location map implementation?

We still need the specialized Ion location map. It accounts for the majority of the performance improvement in the cases when we're filling location metadata.

**Comment by jobarr-amzn** [comment]
> > If we _can_ use `kotlinx.collections.immutables.ImmutableList`, etc. then I'd rather do that. Not because we don't have a perfectly good immutable list implementation in this PR, but because I'd rather own less code.
> > Do we still need or benefit from our own specialized Ion location map implementation?
> 
> We still need the specialized Ion location map. It accounts for the majority of the performance improvement in the cases when we're filling location metadata.

That's what I thought. So if we use `kotlinx.collections.immutables` then we're down to owning only that?

I'm finding documentation on `kotlinx.collections.immutables` a little thin on the ground. The kotlinx immutables library proposal [Use Cases](https://github.com/Kotlin/kotlinx.collections.immutable/blob/master/proposal.md#use-cases) include:
> * Avoiding defensive copying of collection passed as a parameter when you need an unchangeable snapshot and you're unsure whether the collection could be changed later.

While the [API details] section includes:
> Immutable collections just extend the read-only collections. The immutability is enforced solely by their contract: the implementors are required to keep the collection elements unchanged after the collection is constructed.

Then the implication is that [`toImmutableList/Set/Map`](https://github.com/Kotlin/kotlinx.collections.immutable/blob/master/proposal.md#toimmutablelistsetmap) don't perform defensive copies as you said. 

Checking the implementation it looks like this behavior comes from creating an empty immutable thing and then adding the other elements to it while still using the storage of source collection, as described in the README [`+` and `-` operators](https://github.com/Kotlin/kotlinx.collections.immutable/tree/master?tab=readme-ov-file#-and---operators). Is that right?

**Comment by popematt** [comment]
> That's what I thought. So if we use kotlinx.collections.immutables then we're down to owning only that?

We can eliminate the interfaces and collections I added, aside from the special IonLocation map.

> Then the implication is that [toImmutableList/Set/Map](https://github.com/Kotlin/kotlinx.collections.immutable/blob/master/proposal.md#toimmutablelistsetmap) don't perform defensive copies as you said.
>
> Checking the implementation it looks like this behavior comes from creating an empty immutable thing and then adding the other elements to it while still using the storage of source collection, as described in the README [+ and - operators](https://github.com/Kotlin/kotlinx.collections.immutable/tree/master?tab=readme-ov-file#-and---operators). Is that right?

That's what I would have hoped. However, [`toImmutableList()`](https://github.com/Kotlin/kotlinx.collections.immutable/blob/75ed654e34e7111a076d87b8c075fd140e6b4cf7/core/commonMain/src/extensions.kt#L574) actually returns a `PersistentList` (which, to be fair, is an implementation of `ImmutableList`) by delegating to [`toPersistentList()`](https://github.com/Kotlin/kotlinx.collections.immutable/blob/75ed654e34e7111a076d87b8c075fd140e6b4cf7/core/commonMain/src/extensions.kt#L598)... which creates the `PersistentList` by creating an empty persistent list and then adding all of the elements to that persistent list (`persistentListOf<T>() + this`). Then, we see that [`operator fun <E> PersistentList<E>.plus(elements: Iterable<E>)`](https://github.com/Kotlin/kotlinx.collections.immutable/blob/75ed654e34e7111a076d87b8c075fd140e6b4cf7/core/commonMain/src/extensions.kt#L156) delegates to [`PersistentList.addAll(elements: Collection<E>)`](https://github.com/Kotlin/kotlinx.collections.immutable/blob/75ed654e34e7111a076d87b8c075fd140e6b4cf7/core/commonMain/src/ImmutableList.kt#L76). Since we're adding to an empty persistent list, the actual implementation is going to be [`SmallPersistentVector.addAll(elements: Collection<E>)`](https://github.com/Kotlin/kotlinx.collections.immutable/blob/75ed654e34e7111a076d87b8c075fd140e6b4cf7/core/commonMain/src/implementations/immutableList/SmallPersistentVector.kt#L38). At this point, there are two possible execution paths. If there are less than 32 elements being added, it will allocate a new array to hold the elements, copy them into the array, and then returns a new `SmallPersistentVector` that wraps the new array. If there are more than 32 elements, it's even worse—it instead delegates to [`PersistentList<T>.mutate(mutator: (MutableList<T>) -> Unit)`](https://github.com/Kotlin/kotlinx.collections.immutable/blob/75ed654e34e7111a076d87b8c075fd140e6b4cf7/core/commonMain/src/extensions.kt#L41). 

That function calls [`SmallPersistentVector.builder()`](https://github.com/Kotlin/kotlinx.collections.immutable/blob/75ed654e34e7111a076d87b8c075fd140e6b4cf7/core/commonMain/src/implementations/immutableList/SmallPersistentVector.kt#L126), which creates a new instance of [`PersistentVectorBuilder`](https://github.com/Kotlin/kotlinx.collections.immutable/blob/75ed654e34e7111a076d87b8c075fd140e6b4cf7/core/commonMain/src/implementations/immutableList/PersistentVectorBuilder.kt#L15-L25) (which has a decent number of auxiliary fields there). Then it calls [`PersistentVectorBuilder.addAll(elements: Collection<E>)`](https://github.com/Kotlin/kotlinx.collections.immutable/blob/75ed654e34e7111a076d87b8c075fd140e6b4cf7/core/commonMain/src/implementations/immutableList/PersistentVectorBuilder.kt#L160) which starts creating an array of arrays, I think, copying in all of the elements from the collection. Finally, it goes to [`PersistentVectorBuilder.build()`](https://github.com/Kotlin/kotlinx.collections.immutable/blob/75ed654e34e7111a076d87b8c075fd140e6b4cf7/core/commonMain/src/implementations/immutableList/PersistentVectorBuilder.kt#L29). At this point, it creates a `PersistentVector` (since there was more than 32 elements).

----

That is what I was trying to avoid. However, I did end up finding [`ImmutableListAdapter`](https://github.com/Kotlin/kotlinx.collections.immutable/blob/75ed654e34e7111a076d87b8c075fd140e6b4cf7/core/commonMain/src/adapters/ReadOnlyCollectionAdapters.kt#L23) (and its friends), which will actually do what we want them to do.

**Review by jobarr-amzn — APPROVED** [review]

**Pull request merged**